### PR TITLE
nixos/gollum: add options for user and group of gollum process

### PR DESCRIPTION
**Pull request by jooooscha**
###### Description of changes

I added two options: user and group. Both are applied to the gollum user/process.
This is required if the folder of the wiki is a folder of which the current user needs to have ownership.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
I think it would be good to allocate static UID if you go this way: https://github.com/NixOS/nixpkgs/blob/08721b74cafca11ea82e641f19ed1889becaeb2e/nixos/modules/misc/ids.nix

Otherwise, do you think `DynamicUser` could work on that module (or irrelevant?)?

**Comment by jooooscha** [comment]
I am not sure why both changes would be necessary or what they would change.

But I am not that experience with them, either. So if you think it were better to change them, I will certainly do that

**Comment by RaitoBezarius** [comment]
> I am not sure why both changes would be necessary or what they would change.
> 
> But I am not that experience with them, either. So if you think it were better to change them, I will certainly do that

`DynamicUser` is a systemd feature that allocates a dynamic user and manage the UID/GIDs allocations for you, making the first change useless. 

The first change allocates a static UID/GID to prevent any clash, i.e. two services sharing the same UIDs/GIDs (due to recycling or whatever) which would let the other service access Gollum data.

I would advise to go for the first change and we may revisit it later through DynamicUser if possible at all.

**Comment by jooooscha** [comment]
That does not sound like the gollum service would have the same uid as me (the user).
The whole point of this PR is to be able to give the gollum service the same uid/user than the user itself, so I can have the wiki folder in my home directy with me as the owner.

**Comment by RaitoBezarius** [comment]
> That does not sound like the gollum service would have the same uid as me (the user). The whole point of this PR is to be able to give the gollum service the same uid/user than the user itself, so I can have the wiki folder in my home directy with me as the owner.

Hm, in that case, I don't think this PR is right. You are still creating a `gollum` account.

I suggest to have a `mkIf (cfg.user == "gollum")` and create conditionally the account if needed.

**Comment by jooooscha** [comment]
The option for the user is named gollum, but as far as I know, the `name` options overrides this. Therefore, If I set this to my actual linux user account, no extra account would be added. But I could also just rename the `gollum` account, If, for example, I wanted the extra account but with a different name.

**Comment by RaitoBezarius** [comment]
> The option for the user is named gollum, but as far as I know, the `name` options overrides this. Therefore, If I set this to my actual linux user account, no extra account would be added. But I could also just rename the `gollum` account, If, for example, I wanted the extra account but with a different name.
> 
> Maybe we should make this clearer by not using `gollum`, but I don't see a logical error here.

I am sorry I was not clear enough.

Precisely, in nixpkgs, we have patterns like this: https://cs.github.com/NixOS/nixpkgs/blob/e38618a6571c02f9dca1d06189e6d506e4aafe5c/nixos/modules/services/misc/ombi.nix?q=mkIf+%28cfg.user#L71-L79

It is easier to read because you do not have to know the implementation details of `name = xxx` in the user module.

If you can adapt this pattern to gollum, I think it will achieve what you want to do and can be merged.

**Comment by jooooscha** [comment]
Yes, I get your point. However, your proposed change, would prevent the use of any other username for the service, other than `gollum` or the own Linux user. But, I don't care. If you think that is not needed, I am ok with that

**Comment by RaitoBezarius** [comment]
> Yes, I get your point. However, your proposed change, would prevent the use of any other username for the service, other than `gollum` or the own Linux user. But, I don't care. If you think that is not needed, I am ok with that
> 
> Anyway, thanks for your assistance and taking your time :)

Non-standard users are still possible, as long as the operator creates them himself rather than relying on the module doing everything, this is on purpose.

Thank you for the change!

File: nixos/modules/services/misc/gollum.nix [hunk 1]
```diff
@@ -91,18 +91,32 @@ in
...
   };
 
   config = mkIf cfg.enable {
 
-    users.users.gollum = {
-      group = config.users.users.gollum.name;
-      description = "Gollum user";
-      createHome = false;
-      isSystemUser = true;
+    users.users = mkIf (cfg.user == "gollum") {
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/gollum.nix:110]: formatting: `users.users.gollum = mkIf (cfg.user == "gollum")`

**Comment by RaitoBezarius** [comment]
The systemd unit needs changes on `User/Group` to use `cfg.user/cfg.group` too

**Comment by jooooscha** [comment]
One question, I read, I can automatically create a backport PR by adding the `release-22.11` label. However, I don't seem to have the rights to add labels. Can only admins to this?

**Comment by RaitoBezarius** [comment]
> One question, I read, I can automatically create a backport PR by adding the `release-22.11` label. However, I don't seem to have the rights to add labels. Can only admins to this?

Hm, it may be the case indeed.

**Comment by RaitoBezarius** [comment]
Could I ask you to squash the commits and prefix it with `nixos/gollum: <add option for user and group options>` ?

File: nixos/modules/services/misc/gollum.nix [hunk 2]
```diff
@@ -91,18 +91,30 @@ in
...
+      default = "gollum";
+      description = lib.mdDoc "Specifies the owner group of the wiki directory";
+    };
   };
 
   config = mkIf cfg.enable {
 
-    users.users.gollum = {
-      group = config.users.users.gollum.name;
+    users.users.gollum = mkIf (cfg.user == "gollum") {
```
[the remark this answers is not in the document]
Comment (jooooscha) [reply nixos/modules/services/misc/gollum.nix:110]: I wanted to always create a user, originally, but @RaitoBezarius said I should change it.

This is my first PR for nix packages, I don't know your coding style or preferred defaults. Therefore, I just followed his suggestion.

However, I think your idea is much better. I will change it back, as soon as I have time.

**Comment by RaitoBezarius** [comment]
My point was this is more or less a standard over all in nixpkgs. Can you
be more explicit about your usecases and add a comment to explain why to do
something different than usual (in the code) ?

File: nixos/modules/services/misc/gollum.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (erictapen) [reply nixos/modules/services/misc/gollum.nix:110]: My apologies, I wasn't aware this is a common pattern as @RaitoBezarius pointed out. All good then!

**Comment by jooooscha** [comment]
Any update on this? Can this be merged?

**Review by erictapen — APPROVED** [review]
I think I forgot to approve this back then. Thank you!

Interested in becoming a maintainer for the gollum module/package?

**Comment by erictapen** [comment]
> Sure. Why not!

Cool, care to add another PR to add yourself as maintainer?

**Comment by jooooscha** [comment]
Done: #241088

**Pull request merged**